src/snowflake/snowpark/_internal/analyzer/analyzer.py

```diff
@@ -166,18 +167,33 @@ def __init__(self, session: "snowflake.snowpark.session.Session") -> None:
         self.generated_alias_maps = {}
         self.subquery_plans = []
         self.alias_maps_to_use: Optional[Dict[uuid.UUID, str]] = None
+        self.config_context: ConfigContext = ConfigContext(session)
+        # Point this plan builder to the same config context as the analyzer
+        self.plan_builder.config_context = self.config_context
 
     def analyze(
         self,
         expr: Union[Expression, NamedExpression],
         df_aliased_col_name_to_real_col_name: DefaultDict[str, Dict[str, str]],
         parse_local_name=False,
+    ) -> str:
+        # Set the config context for analysis step
+        with self.config_context:
+            return self.do_analyze(
+                expr, df_aliased_col_name_to_real_col_name, parse_local_name
+            )
+
+    def do_analyze(
+        self,
+        expr: Union[Expression, NamedExpression],
+        df_aliased_col_name_to_real_col_name: DefaultDict[str, Dict[str, str]],
```

Comment:
one way i can think about is to let every function that you want to protect with take a configContext parameter, and let session provide a getConfigSnapshot capability.
in terms code change, i think it is probably similar as what you have here. 

When i look at the problem, the issue seems very easy to address during the new compilation stage, because the code generator is create per compilation. However, for the general analyzer, it is one per session, which makes things little bit harder.

If we don't do any snapshot here, and do not consider about the old cte optimization. is the problem only happens during the plan building for the new compilation? if that is the case, maybe we can handle the new compilation stage, and put a rule in the future that all implementation has to be thread safe.

Reply:
> However, for the general analyzer, it is one per session, which makes things little bit harder.

We have one session per-thread. So the config context snapshot we are taking here is taken in each thread.

> if that is the case, maybe we can handle the new compilation stage, and put a rule in the future that all implementation has to be thread safe.

I do not understand this completely. Do you mean we handle plan builder same way we deal with new compilation stage?